### PR TITLE
abi: add AddressSpace field to Primitive::Pointer 

### LINKED ISSUE #106367: function pointer address space handling (for AVR) is wrong
@eddyb [points out](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Fn.20Ptr.20pointee.20info.20is.20nonsense.3F) that the handling of address space for function pointers (introduced by https://github.com/rust-lang/rust/pull/73270) is fundamentally wrong: `PointeeInfo` can be used for *optimization hints only* and also only becomes relevant way too late during compilation. To properly account for function pointers having a different address space, they need to be added in the `Pointer` variant of [the `Primitive` ABI type](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_abi/enum.Primitive.html). That is the only way to fix types like `Result<&i32, fn()>`, which are currently going horribly wrong for Harvard architectures because the enum layout algorithm will think that `&i32` and `fn()` are exactly the same fundamental type, and put them into one and the same field (resulting LLVM layout: `{ i64, ptr }`).

Not sure whom to ping here, so I will pick some people that seem to be active in an around AVR: @dylanmckay @Patryk27.
Also Cc @rust-lang/compiler since this is related to some pretty fundamental compiler infrastructure

### PR DESCRIPTION
**Pull request by erikdesjardins**
...and remove it from `PointeeInfo`, which isn't meant for this.

There are still various places (marked with FIXMEs) that assume all pointers
have the same size and alignment. Fixing this requires parsing non-default
address spaces in the data layout string (and various other changes),
which will be done in a followup.
(That is, if it's actually worth it to support multiple different pointer sizes.
There is a lot of code that would be affected by that.)

Fixes #106367

r? @oli-obk
cc @Patryk27

### CODE REVIEW

File: compiler/rustc_codegen_gcc/src/consts.rs [hunk 1]
```diff
@@ -322,13 +322,21 @@ pub fn const_alloc_to_gcc<'gcc, 'tcx>(cx: &CodegenCx<'gcc, 'tcx>, alloc: ConstAl
             )
             .expect("const_alloc_to_llvm: could not read relocation pointer")
             as u64;
+
+        let address_space = match cx.tcx.global_alloc(alloc_id) {
+            GlobalAlloc::Function(..) => cx.data_layout().instruction_address_space,
+            GlobalAlloc::Static(..) | GlobalAlloc::Memory(..) | GlobalAlloc::VTable(..) => {
+                AddressSpace::DATA
+            }
+        };
```
Comment (oli-obk) [inline compiler/rustc_codegen_gcc/src/consts.rs:331]: this should probably be a method on `GlobalAlloc` (even if it has to take a `HasDataLayout` argument)

**Review by oli-obk — APPROVED** [review]
looks like that snippet about GlobalAlloc is only ever used once. A bit surprising, but maybe we'll see more of it if we end up teaching const eval about address spaces (it already kind of has some of them implicitly)

File: compiler/rustc_codegen_gcc/src/consts.rs (hunk 1 above)
Comment (erikdesjardins) [reply compiler/rustc_codegen_gcc/src/consts.rs]: This is actually copied from cg_llvm, so it is used in two places... (fixed in latest commit)

File: compiler/rustc_abi/src/lib.rs [hunk 2]
```diff
@@ -1468,7 +1461,6 @@ pub struct PointeeInfo {
     pub size: Size,
     pub align: Align,
     pub safe: Option<PointerKind>,
-    pub address_space: AddressSpace,
```
Comment (RalfJung) [inline compiler/rustc_abi/src/lib.rs:1471]: Awesome, thanks a ton. :)
Note to self: add a comment to `PointeeInfo` stating that this is for optimization purposes only and safe to ignore by backends (as suggested by @eddyb)

**Pull request merged**